src/linear_algebra/matrix/charpoly/eigs.lean

```diff
@@ -0,0 +1,79 @@
+/-
+Copyright (c) 2023 Mohanad Ahmed. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mohanad Ahmed
+-/
+
+import data.polynomial.basic
+import field_theory.is_alg_closed.basic
+import algebra.char_p.two
+
+/-!
+# Eigenvalues are characteristic polynomial roots.
+
+
+In algebraically closed fields, we show:
+* `matrix.det_eq_prod_roots_charpoly`: the determinant is the product of the roots of the characteristic
+polynomial.
+* `matrix.trace_eq_sum_roots_charpoly`: the trace is the sum of the roots of the characteristic
+polynomial.
+
+Note that over other fields such as ` ℝ`, these results can be used by using `A.map (algebra_map ℝ ℂ)` as the matrix,
+and then applying `ring_hom.map_det`.
+-/
+variables {n: Type*}[fintype n][decidable_eq n]
+variables {R: Type*}[field R][is_alg_closed R]
+
+open matrix polynomial
+open linear_map module.End
+open_locale matrix big_operators
+
+lemma prod_roots_charpoly_eq_det (A : matrix n n R) :
+  A.det = (matrix.charpoly A).roots.prod :=
+begin
+  by_cases hn: nonempty n,
+  { have hdeg := charpoly_nat_degree_eq_dim A,
+    rw [det_eq_sign_charpoly_coeff, ← hdeg,
+      polynomial.prod_roots_eq_coeff_zero_of_monic_of_split,
+      ← mul_assoc, ← pow_two, ← pow_mul ],
+
+    by_cases h2: ring_char R ≠ 2,
+    { have hstupid: -1 ≠ (1:R), {exact ring.neg_one_ne_one_of_char_ne_two h2,},
+      have hs2 : even(A.charpoly.nat_degree*2), {simp only [even.mul_left, even_two],},
+      rw [(neg_one_pow_eq_one_iff_even hstupid).2 hs2, one_mul],},
+    { rw [ne.def, not_not] at h2,
+      rw [neg_one_eq_one_iff.2 h2, one_pow, one_mul],},
+
+    apply matrix.charpoly_monic,
+    exact is_alg_closed.splits A.charpoly, },
```

Comment:
Case-splitting on char 2 is usually an indication that you're using the wrong lemmas!
```suggestion
    rw [det_eq_sign_charpoly_coeff, ← hdeg,
      polynomial.prod_roots_eq_coeff_zero_of_monic_of_split A.charpoly_monic
        (is_alg_closed.splits _),
      ← mul_assoc, ← pow_two, pow_right_comm, neg_one_sq, one_pow, one_mul],
```